site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,768 @@
+	// equivalent.
+	Name GRPCHeaderName `json:"name"`
+
+	// Value is the value of the gRPC Header to be matched.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=4096
+	Value string `json:"value"`
+}
+
+// GRPCRequestMirrorFilter defines configuration for the RequestMirror filter.
+type GRPCRequestMirrorFilter struct {
+	// BackendRef references a resource where mirrored requests are sent.
+	//
+	// If the referent cannot be found, this BackendRef is invalid and must be
+	// dropped from the Gateway. The controller must ensure the "ResolvedRefs"
+	// condition on the Route status is set to `status: False` and not configure
+	// this backend in the underlying implementation.
+	//
+	// If there is a cross-namespace reference to an *existing* object
+	// that is not allowed by a ReferencePolicy, the controller must ensure the
+	// "ResolvedRefs"  condition on the Route is set to `status: False`,
+	// with the "RefNotPermitted" reason and not configure this backend in the
+	// underlying implementation.
+	//
+	// In either error case, the Message of the `ResolvedRefs` Condition
+	// should be used to provide more detail about the problem.
+	//
+	// Support: Extended for Kubernetes Service
+	// Support: Custom for any other resource
+	BackendRef BackendObjectReference `json:"backendRef"`
+}
+```
+
+In addition, two new `ProtocolType` options will be added:
+
+```diff
+--- a/apis/v1alpha2/gateway_types.go
++++ b/apis/v1alpha2/gateway_types.go
+@@ -165,6 +165,11 @@ type Listener struct {
+        //   protocol layers as described above. If an implementation does not
+        //   ensure that both the SNI and Host header match the Listener hostname,
+        //   it MUST clearly document that.
++       // * H2C: The Listener Hostname MUST match the :authority pseudo-header of the request.
++       // * H2: The Listener Hostname SHOULD match at both the TLS and HTTP
++       //   protocol layers as described above. If an implementation does not
++       //   ensure that both the SNI and :authority pseudo-header match the Listener hostname,
++       //   it MUST clearly document that.
+        //
+        // For HTTPRoute and TLSRoute resources, there is an interaction with the
+        // `spec.hostnames` array. When both listener and route specify hostnames,
+@@ -279,6 +284,12 @@ const (
+ 
+        // Accepts UDP packets.
+        UDPProtocolType ProtocolType = "UDP"
++
++       // Accepts HTTP/2 sessions over TLS.
++       H2ProtocolType ProtocolType = "H2"
++
++       // Accepts cleartext HTTP/2 sessions.
++       H2CProtocolType ProtocolType = "H2C"
+ )
+ 
+ // GatewayTLSConfig describes a TLS configuration.
+```
+
+TLS-encryption is the default in HTTP/2, so H2C is the
+[marked](https://en.wikipedia.org/wiki/Markedness) option here.
+
+If the listener supports neither `H2` nor `H2C`, then the implementation MUST
+raise an `Accepted` Condition with a status of `False` in the corresponding
+RouteParentStatus.
+
```

Comment:
I know we don't always ask for this, but because this is the first time we're adding a new Route resource, I'd like to see something about the graduation plan.

Based on what we have in #1039, I'd suggest something like:
- gRPC resource implemented by a couple of implementations
- Some conformance tests in place
- Users are using the API surface
- API review done by upstream reviewers.

Graduation to GA should be similar but more stringent (exhaustive conformance tests, implemented by lots o implementations, etc).